### DISCUSSION TITLE
kafka topic naming conventions

### DISCUSSION THREAD

**Discussion by shaiktas** [milvus-io/milvus#36095] [Q&A and General discussion]
Hello,
How to understand kafka tipic and collection mapping? We do see topics  directories as below under /bitname/kafka/data directory as:How to understand which topic is belongs to which collection? Kindly help us to understand these topic name conventions?
milvus-isds-rootcoord-dml_2-0/0
milvus-isds-rootcoord-dml_2-0/0
milvus-isds-rootcoord-dml_6-0/0
milvus-isds-rootcoord-dml_10-0/
milvus-isds-rootcoord-dml_7-0/0
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_2-0/0
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/
milvus-isds-rootcoord-dml_14-0/

**Comment by yhmo** [comment]
In milvus source code, 

Milvus utilizes two types of channels, [PChannel](https://milvus.io/docs/glossary.md#PChannel) and [VChannel](https://milvus.io/docs/glossary.md#VChannel). Each PChannel corresponds to a topic for log storage, while each VChannel corresponds to a shard in a collection.

"PChannel" ------- physical channel
"VChannel" ------- virtual channel

Each kafka topic is a "physical channel". The number of physical channels is defined in the milvus.yaml:
```
rootCoord:
  dmlChannelNum: 16 # The number of DML-Channels to create at the root coord startup.
```
By default, milvus creates 16 physical channels during startup.


When you create a collection, a parameter named "shards_num" defines its virtual channels. 
For example, create a collection with 2 virtual channels:
`collection = Collection(name = collection_name, schema = schema, shards_num = 2)`

Each time you insert a batch of data, the data is split into N pieces and stored into kafka/pulsar. Then querynodes/datanodes will consume the pieces of data from "virtual channels".

Multiple vchannels can share one pchannel. 
A P-channel does not always serve one collection. V-channel is assigned to a P-channel by a runtime balancer. A collection with multiple vchannels, the multiple vchannels could be assigned to multiple pchannels, or assigned to a single P-channel, depending on the runtime balancer.

The diagram can help understanding the relationship between pchannel and vchannel:
![channels drawio](https://github.com/user-attachments/assets/d46e3cd7-6be9-4d4e-b7a2-f03821c9cebe)

**Comment by shaiktas** [comment]
Thanks @yhmo . It helps me alot. So i understood the milvus-kafka wil have max 16 topics and will locate all streaming data from all collections within these 16 topics. Assume if data ingestion is too high, is it good practive to increase these config to higher value ie 32 so that it can create 32 topics to locate the data and can stream faster from kafka to storage (s3).  Because we are seeing  heavy date `.log` files of size 1 GB in kafka brokers under /bitnami/kafka/data/. Why is this limitted to 1GB?? any configs driving this file size?
1.0G Sep  2 19:32 milvus-xxxxx-rootcoord-dml_8-0/00000000000020126342.log

**Reply by yhmo** [reply]
Yes, you can increase the rootCoord.dmlChannelNum value if you have many collections with continually streaming data inserts.
The kafaka is an independent service, I think the kafka service, not milvus, configures in the size of log file.

**Comment by shaiktas** [comment]
Got it @yhmo ,  due to some reason we see data nodes are crashing with OOMKill Reason. After we clean up kafka  (Replaced fresh kafka external cluster from milvus yaml) ,  data nodes back and stable. Up on investigation we found that most of the topics in kafka brokers  (/bitnake/kafka/data/by-dev-rootcoord-dml-*/*.log are 1GB and observed 70+ log files . The other side we have data nodes (kafka consumer) with 5 as replica and each data node is of 64GB RAM.  Data nodes cpu and memory is not exceeded. Any pointers or direction how to debug this issue?

**Reply by yhmo** [reply]
Theoretically, the memory usage of data node has nothing to do with pulsar/kafka because the data is consumed in streaming.

Data nodes mainly do the following things:
1) Subscribes data from kafka/pulsar, and maintains an in-memory buffer for each shard. Data is accumulated in the buffer. A growing segment for each shard. Once the buffer size reaches a threshold(default 16MB), data node will persist the buffer as a blob file in S3. Each segment has multiple blob files. Once the total size of the growing segment's blob files reaches a threshold, the growing segment is turned into a sealed segment.
2) Checks segments to see if there are small segments that can be compacted. Each data node only checks segments of the shard that it is maintaining. If there are segments that can be compacted, the data node will begin a compaction task in the background. Data node reads the segments into memory blob by blob, and merges the data in memory, persists into a new blob for the new large segment.


The memory usage of (1) is controlled by these configurations in the milvus.yaml:
The forceSyncWatermark is 0.5, which means the total size of in-memory buffers are controlled under 50% of the memory capacity.
```
dataNode:
  memory:
    forceSyncEnable: true # Set true to force sync if memory usage is too high
    forceSyncSegmentNum: 1 # number of segments to sync, segments with top largest buffer will be synced.
    checkInterval: 3000 # the interal to check datanode memory usage, in milliseconds
    forceSyncWatermark: 0.5 # memory watermark for standalone, upon reaching this watermark, segments will be synced.
```
The memory usage of (2) is not high because it only reads a few blobs instead of loading the entire segment.

To get more information, you can deploy a monitoring system for milvus, and observe the metrics of data node:
https://milvus.io/docs/visualize.md

Some known issues that cause data node OOM, some of them have been fixed in the latest versions:
https://github.com/milvus-io/milvus/issues/34703
https://github.com/milvus-io/milvus/issues/33676
https://github.com/milvus-io/milvus/issues/33547

**Comment by shaiktas** [comment]
Thanks for explanation @yhmo . We understood that milvus will publishes data to 16 channels (as per rootCoord:  dmlChannelNum=16) config. How about consumers for these topics/channels? Assume if there are 5 data nodes (replicas:5). does it mean that only 5 consumers for 16 topics? Do we have any mappings between consumer and topics? As per kafka it is recommended to have one consumer per topic partition .

**Reply by yhmo** [reply]
Each virtual channel will create a consumer.
If a collection's shards_num = 2, the collection has 5 partitions, the total vchannel in this collection is 2 * 5 = 10, there will be 10 consumers for this collection.

**Reply by shaiktas** [reply]
Thank you @yhmo for sharing insights. As per my understanding only datanodes/index nodes acts as streaming (kafka) consumer, If there are 10 vchannels, there will be 10 consumers , so all these 10 consumers are from within existing data nodes? assume i have a 3 data nodes running. How consumers will get split per data node?

**Reply by yhmo** [reply]
Actually, querynodes/datanodes consume data from pulsar/kafka.
querynodes consume data for search/query if collection is loaded.
datanodes consume data for persistence.

V-channels are balanced between datanodes and querynodes.
Let's say there are total of 100 v-channels. 3 data nodes, ideally there will be 33 ~ 34 consumers per data node. 4 query nodes, there will be 25 consumers per query node.

If there is only one v-channel, then only one data node consumes data from kafka, and only one query node consumes data from kafka.
